spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -2458,17 +2472,29 @@ private void handleNack(final ConsumerRecords<K, V> records, final ConsumerRecor
 				}
 			}
 			SeekUtils.doSeeks(list, this.consumer, null, true, (rec, ex) -> false, this.logger); // NOSONAR
-			nackSleepAndReset();
+			pauseForNackSleep();
 		}
 
-		private void nackSleepAndReset() {
-			try {
-				ListenerUtils.stoppableSleep(KafkaMessageListenerContainer.this.thisOrParentContainer, this.nackSleep);
-			}
-			catch (@SuppressWarnings(UNUSED) InterruptedException e) {
-				Thread.currentThread().interrupt();
+		private void pauseForNackSleep() {
+			if (this.nackSleep > 0) {
+				this.nackWake = System.currentTimeMillis() + this.nackSleep;
+				this.nackSleep = -1;
+				Set<TopicPartition> alreadyPaused = this.consumer.paused();
+				this.pausedForNack.addAll(getAssignedPartitions());
+				this.pausedForNack.removeAll(alreadyPaused);
+				this.logger.debug(() -> "Pausing for nack sleep: " + ListenerConsumer.this.pausedForNack);
+				try {
+					this.consumer.pause(this.pausedForNack);
+				}
+				catch (IllegalStateException ex) {
+					// this should never happen; defensive, just in case...
+					this.logger.warn(() -> "Could not pause for nack, possible rebalance in process: "
+							+ ex.getMessage());
+					Set<TopicPartition> nowPaused = new HashSet<>(this.consumer.paused());
+					nowPaused.removeAll(alreadyPaused);
+					this.consumer.resume(nowPaused);
```

Comment:
Can we expect `consumer.resume()` to succeed if `consumer.pause()` threw an IllegalStateException?

Reply:
`pause()` iterates over the collection and throws an exception if the partition is no longer assigned; therefore some partitions may have been paused; all this code is doing is resuming those that were successful. As the comment states, this code should never be executed. I added it before I removed the paused (for nack) partitions in `onPartitionsRevoked()`, but decided to leave it here (with the comment).

Reply:
I see, thanks for the blazing fast reaction on the bug report and the detailed explanation above!